src/stripeLogsView.ts

```diff
@@ -1,15 +1,211 @@
 export class StripeLogsDataProvider extends StripeTreeViewDataProvider {
+  private static readonly REFRESH_DEBOUNCE_MILLIS = 1000;
+
+  private stripeClient: StripeClient;
+  private logTreeItems: StripeTreeItem[];
+  private viewState: ViewState;
+  private logsStdoutStream: stream.Writable | null;
+  private logsStderrStream: stream.Writable | null;
+
+  constructor(stripeClient: StripeClient) {
+    super();
+    this.stripeClient = stripeClient;
+    this.logTreeItems = [];
+    this.logsStdoutStream = null;
+    this.logsStderrStream = null;
+    this.viewState = ViewState.Idle;
+  }
+
+  startLogsStreaming = async () => {
+    try {
+      await this.setupStreams();
+    } catch (e) {
+      window.showErrorMessage(e.message);
+      this.stopLogsStreaming();
+    }
+  };
+
+  stopLogsStreaming = () => {
+    this.cleanupStreams();
+    this.setViewState(ViewState.Idle);
+  };
+
   buildTree(): Promise<StripeTreeItem[]> {
-    const logStreamItem = new StripeTreeItem('Start API logs streaming', 'openLogsStreaming');
-    logStreamItem.setIcon({
-      dark: Resource.ICONS.dark.terminal,
-      light: Resource.ICONS.light.terminal,
-    });
+    const streamingControlItemArgs = (() => {
+      switch (this.viewState) {
+        case ViewState.Idle:
+          return {
+            label: 'Start streaming API logs',
+            command: 'startLogsStreaming',
+            iconId: 'play-circle',
+          };
+        case ViewState.Loading:
+          return {
+            label: 'Starting streaming API logs...',
+            command: 'stopLogsStreaming',
+            iconId: 'loading',
+          };
+        case ViewState.Streaming:
+          return {
+            label: 'Stop streaming API logs',
+            command: 'stopLogsStreaming',
+            iconId: 'stop-circle',
+          };
+      }
+    })();
+
+    const streamingControlItem = this.createItemWithCommand(streamingControlItemArgs);
+
+    const treeItems = [streamingControlItem];
+
+    if (this.logTreeItems.length > 0) {
+      const logsStreamRootItem = new StripeTreeItem('Recent logs');
+      logsStreamRootItem.children = this.logTreeItems;
+      logsStreamRootItem.expand();
+      treeItems.push(logsStreamRootItem);
+    }
+
+    return Promise.resolve(treeItems);
+  }
+
+  private createItemWithCommand({
+    label,
+    command,
+    iconId,
+  }: {
+    label: string;
+    command?: string;
+    iconId?: string;
+  }) {
+    const item = new StripeTreeItem(label, command);
+    if (iconId) {
+      item.iconPath = new ThemeIcon(iconId);
+    }
+    return item;
+  }
+
+  private async setupStreams() {
+    this.setViewState(ViewState.Loading);
+
+    const stripeLogsTailProcess = await this.stripeClient.getOrCreateCLIProcess(
+      CLICommand.LogsTail,
+      ['--format', 'JSON'],
+    );
+    if (!stripeLogsTailProcess) {
+      throw new Error('Failed to start `stripe logs tail` process');
+    }
+
+    stripeLogsTailProcess.on('close', this.stopLogsStreaming);
```

Comment:
can `setupStreams` be called more than once? It's not a huge leak, but if it does, this will continually append the handler to the set of listeners.

Reply:
Yeah `setupStreams` could be called more than once, so I added a conditional to the caller of `setupStreams`.